lib/BaseController.js

```diff
@@ -108,42 +108,107 @@ module.exports = class BaseController {
+      if (objectPath.get(this._data, ['object', 'status', 'deploy.razee.io/finalizer-cleanup']) == 'running') {
+        this._logger.debug(`Found deletionTimestamp.. but finalizer already running.. skipping ${this._data.type} event ${objectPath.get(this._data, 'object.metadata.selfLink')} ${objectPath.get(this._data, 'object.metadata.resourceVersion')}`);
+        return;
+      }
+      this._logger.debug(`Found deletionTimestamp.. running finalizer.. ${objectPath.get(this._data, 'object.metadata.resourceVersion')}`);
+      this._logger.debug(`'Modified' Finalizer ${objectPath.get(this._data, 'object.metadata.selfLink')} started`);
+      let hasDeletionTimestamp = await this.finalizer();
+      this._logger.debug(`'Modified' Finalizer ${objectPath.get(this._data, 'object.metadata.selfLink')} completed: deletionTimestamp ${hasDeletionTimestamp}`);
+      if (objectPath.get(this._data, 'object.metadata.finalizers', []).length > 0) {
+        try {
+          await this._patchStatus(); // to be removed
+          return await this._patchRazeeLogs();
+        } catch (e) {
+          return (e.statusCode === 404) ? { message: 'Resource already deleted', statusCode: 404 } : Promise.reject(e);
+        }
+      }
+      return;
+    }
+
+    // if data, deemed important, has changed (identified via the data hash), modified() should run
+    let dh = objectPath.get(this._data, ['object', 'metadata', 'annotations', 'deploy.razee.io/data-hash']);
+    let cdh = this.computeDataHash(objectPath.get(this._data, 'object'));
+    if (dh != cdh) {
+      this._logger.debug(`Last known deploy.razee.io/data-hash doesn't match computed.. updating annotation and running modified().. ${objectPath.get(this._data, 'object.metadata.resourceVersion')}`);
+      await this.patchSelf({ metadata: { annotations: { 'deploy.razee.io/data-hash': cdh } } });
+      await this.modified();
+    } else { // else non significant change has occured, event is skipped
+      this._logger.info(`No relevant change detected.. skipping ${this._data.type} event ${objectPath.get(this._data, 'object.metadata.selfLink')} ${objectPath.get(this._data, 'object.metadata.resourceVersion')}`);
+    }
   }
   async modified() {
-    await this._added();
+    return await this._added();
   }
 
   async _deleted() {
-    this.deleted();
+    return await this.deleted();
   }
   async deleted() {
-    this._logger.info(this._data);
+    return this._logger.info(this._data);
+  }
+
+  // General helpers ===========================================
+  async _cluster_locked() {
+    let lockCluster = 'false';
+    let lockClusterPath = './config/lock-cluster';
+    let exists = await fs.pathExists(lockClusterPath);
+    if (exists) {
+      lockCluster = await fs.readFile(lockClusterPath, 'utf8');
+      lockCluster = lockCluster.trim().toLowerCase();
+    }
+    return (lockCluster == 'true');
+  }
+
+  computeDataHash(resource) {
```

Comment:
Refactor to make it simpler for subclasses to identify `importantData`

Reply:
done